dali/base/dasds.cpp

```diff
@@ -2095,6 +2101,146 @@ ISDSManagerServer &querySDSServer()
 
 /////////////////
 
+void CExternalFile::remove(const char *name)
+{
+    StringBuffer filename(name);
+    filename.append(ext);
+    manager.extCache.remove(filename);
+    manager.deltaWriter.removeExt(filename.detach());
+}
+
+void CBinaryFileExternal::readValue(const char *name, MemoryBuffer &mb)
+{
+    StringBuffer extName;
+    getName(extName, name);
+    if (manager.extCache.lookup(extName, mb))
+        return;
+
+    StringBuffer filename;
+    getFilename(filename, name);
+    Owned<IFile> iFile = createIFile(filename.str());
+    size32_t sz = (size32_t)iFile->size();
+    if ((unsigned)-1 == sz)
+    {
+        StringBuffer s("Missing external file ");
+        Owned<IException> e = MakeSDSException(SDSExcpt_MissingExternalFile, "%s", filename.str());
+        LOG(MCoperatorWarning, unknownJob, e, s.str());
+        StringBuffer str("EXTERNAL BINARY FILE: \"");
+        str.append(filename.str()).append("\" MISSING");
+        CPTValue v(str.length()+1, str.str(), false);
+        v.serialize(mb);
+    }
+    else
+    {
+        Owned<IFileIO> fileIO = iFile->open(IFOread);
+        verifyex(sz == ::read(fileIO, 0, sz, mb));
+        manager.extCache.add(extName, sz, mb.toByteArray());
+    }
+}
+
+void CBinaryFileExternal::read(const char *name, IPropertyTree &owner, MemoryBuffer &mb, bool withValue)
+{
+    const char *_name = owner.queryName();
+    if (!_name) _name = "";
+    mb.append(_name);
+
+    size32_t flagsPos = mb.length();
+    byte flags = ((PTree &)owner).queryFlags();
+    mb.append(flags);
+    serializeVisibleAttributes(owner, mb);
+    StringBuffer extName;
+    getName(extName, name);
+    if (manager.extCache.lookup(extName, mb))
+        return;
+
+    if (withValue)
+    {
+        StringBuffer filename;
+        getFilename(filename, name);
+        Owned<IFile> iFile = createIFile(filename.str());
+        size32_t sz = (size32_t)iFile->size();
+        if ((unsigned)-1 == sz)
+        {
+            IptFlagClr(flags, ipt_binary);
+            mb.writeDirect(flagsPos, sizeof(flags), &flags);
+
+            StringBuffer s("Missing external file ");
+            if (*_name)
+                s.append("in property ").append(_name);
+            Owned<IException> e = MakeSDSException(SDSExcpt_MissingExternalFile, "%s", filename.str());
+            LOG(MCoperatorWarning, unknownJob, e, s.str());
+            StringBuffer str("EXTERNAL BINARY FILE: \"");
+            str.append(filename.str()).append("\" MISSING");
+            CPTValue v(str.length()+1, str.str(), false);
+            v.serialize(mb);
+        }
+        else
+        {
+            Owned<IFileIO> fileIO = iFile->open(IFOread);
+            verifyex(sz == ::read(fileIO, 0, sz, mb));
```

Comment:
read function unnecessarily checks size() again.  I don't know if that will cause an extra delay (on remote storage), but may be worth avoiding.

Reply:
looks like it would at least be worth modifying ::read to avoid the check:
```
    const size32_t checkLengthLimit = 0x1000;
    if (len >= checkLengthLimit)
    {
        //Don't allocate a stupid amount of memory....
        offset_t fileLength = in->size();
        if (pos > fileLength)
            pos = fileLength;
        if ((len == (size32_t)-1) || (pos + len > fileLength))
            len = (size32_t)(fileLength - pos);
    }
```

if pos == 0 and len != -1

Reply:
make sense to do in separate JIRA/PR : https://track.hpccsystems.com/browse/HPCC-30193